File: PRJ01/src/toggleLED.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#define GPIO_LED "17"
#define GPIO_BUTTON "27"
#define GPIO_LED_PATH "/sys/class/gpio/gpio17/"
#define GPIO_BUTTON_PATH "/sys/class/gpio/gpio27/"
#define GPIO_SYSFS "/sys/class/gpio/"

void writeGPIO(char filename[], char value[])
{
    FILE *file_pointer;
    file_pointer = fopen(filename, "w+");
    fprintf(file_pointer, "%s", value);
    fclose(file_pointer);
}

int main(int argc, char *argv[])
{
    printf("\nStarting the toggleLED program\n");
    printf("----------------------------\n\n");
    usleep(1000000);

    printf("Setting up the LED on the GPIO\n");
    writeGPIO(GPIO_SYSFS "export", GPIO_LED);
    usleep(100000);
    writeGPIO(GPIO_LED_PATH "direction", "out");

    usleep(1000000);

    printf("Setting up the BUTTON on the GPIO\n\n");
    writeGPIO(GPIO_SYSFS "export", GPIO_BUTTON);
    usleep(100000);
    writeGPIO(GPIO_BUTTON_PATH "direction", "in");

    int counter = 0;

    while (counter < 100)
    {
        usleep(100000);

        FILE *file_pointer;
        char line[80], full_file_name[100];
        sprintf(full_file_name, GPIO_BUTTON_PATH "/value");
        file_pointer = fopen(full_file_name, "r");
        
        while (fgets(line, 80, file_pointer) != NULL)
        {
            printf("The state of the BUTTON is %s", line);
        }

        if (line[0] == '1')
        {
            writeGPIO(GPIO_LED_PATH "value", "1");
        }
        else 
        {
            writeGPIO(GPIO_LED_PATH "value", "0");
        }
        fclose(file_pointer);
        counter++;
    }

    usleep(1000000);
    printf("\nTurning the LED off\n");
    writeGPIO(GPIO_LED_PATH "value", "0");

    usleep(1000000);
    printf("Closing the LED on the GPIO\n");
    writeGPIO(GPIO_SYSFS "unexport", GPIO_LED);

    usleep(1000000);
    printf("Closing the BUTTON on the GPIO\n");
    writeGPIO(GPIO_SYSFS "unexport", GPIO_BUTTON);
    
    printf("\n----------------------------\n");
    printf("Finished the toggleLED Program\n\n");
    exit(EXIT_SUCCESS);
}

File: PRJ02/src/toggleLED.py
import os
import time

GPIO_LED = "17"
GPIO_BUTTON = "27"
GPIO_SYSFS = "/sys/class/gpio/"
GPIO_LED_PATH = os.path.join(GPIO_SYSFS, "gpio"+GPIO_LED)
GPIO_BUTTON_PATH = os.path.join(GPIO_SYSFS, "gpio"+GPIO_BUTTON)

def write_GPIO(filename, value):
    with open(filename, "w") as file:
        file.write(value)

def read_GPIO(filename):
    with open(filename) as file:
        content = file.read()
        return content

def main():
    print("\nStarting the toggleLED program\n")
    print("----------------------------\n")
    time.sleep(1)

    print("Setting up the LED on the GPIO\n")
    write_GPIO(GPIO_SYSFS+"export", GPIO_LED)
    time.sleep(0.1)
    write_GPIO(GPIO_LED_PATH+"/direction", "out")

    time.sleep(1)

    print("Setting up the BUTTON on the GPIO\n\n")
    write_GPIO(GPIO_SYSFS+"export", GPIO_BUTTON)
    time.sleep(0.1)
    write_GPIO(GPIO_BUTTON_PATH+"/direction", "in")
    time.sleep(1)

    print("-- Toggle LED with push button or close the program: ctrl+c --")
    
    while True:
        content = read_GPIO(GPIO_BUTTON_PATH+"/value")

        if "1" in content:
            write_GPIO(GPIO_LED_PATH+"/value", "1")
        else:
            write_GPIO(GPIO_LED_PATH+"/value", "0")

        time.sleep(0.1)

try:
    main()
except KeyboardInterrupt:
    time.sleep(1)
    print("\n\nClosing the LED on the GPIO\n")
    write_GPIO(GPIO_SYSFS+"unexport", GPIO_LED)
    time.sleep(0.5)
    print("Closing the BUTTON on the GPIO\n")
    write_GPIO(GPIO_SYSFS+"unexport", GPIO_BUTTON)
    time.sleep(1)
    print("\n----------------------------")
    print("Finished the toggleLED Program\n\n")
except:
    print("\n\nOther error occured!")
    time.sleep(1)
    print("Closing the LED on the GPIO\n")
    write_GPIO(GPIO_SYSFS+"unexport", GPIO_LED)
    time.sleep(0.5)
    print("Closing the BUTTON on the GPIO\n")
    write_GPIO(GPIO_SYSFS+"unexport", GPIO_BUTTON)
    time.sleep(1)
    print("\n----------------------------")
    print("Finished the toggleLED Program\n\n")